### DISCUSSION TITLE
Traffic display

### DISCUSSION THREAD

**Discussion by JanneMantyharju** [makerplane/pyEfis#250] [General]
I have Stratux running on the same device with the pyEfis. I was wondering what would be the best way to implement the traffic instrument. A quick and dirty solution might be to grab the image from the Stratux web-UI and display that. Proper way would be to write a native instrument that uses data provided by Stratux. There seems to be at least few stand-alone projects doing this which code might be adapted to this use.
Is anybody working with this or thought about the solution?

**Comment by e100** [comment]
I added the Android display feature and use Ifly EFB. One could also use Avare that is free.
How to install Android:
https://github.com/makerplane/pyEfis/blob/master/ANDROID.md
How to display Android in pyefis:
https://github.com/makerplane/pyEfis/blob/master/docs/screenbuilder.md#weston

I don't think it would be too difficult to clone and modify the waydroid service monitor to display any linux application, that can natively display on Wayland, within pyefis. Might be possible to make this work for X11 applications too but it's not something I've tried so I don't know how or if it is actually possible. 
Referring to this service:
https://github.com/makerplane/pyEfis/blob/master/ANDROID.md#install-required-services
That runs this script that would need cloned and modified to run and monitor the other application and alternate weston instance instead of waydroid:
https://github.com/makerplane/pyEfis/blob/master/extras/extras/waydroid_monitor.sh
For example, one could run a browser that opens the Stratux page that displays the traffic. 

I just don't think it is worthwhile to develop a feature that is already implemented so well in other applications. But it might be worthwhile to make it easier to display other applications in pyefis.

If one wanted to implement this in pyefis I think adding it to pyavmap would be the best place to do that.  https://github.com/makerplane/pyAvMap/tree/master
We would also need add an option to screenbuilder to display pyavmap.

**Comment by JanneMantyharju** [comment]
I decided to pursue the lowest hanging fruit. I wrote an instrument that contains QtWebView that opens the Stratux traffix page that runs locally on the same computer. This works, but cleaner solution would be to write dedicated traffic page to better suit the look and feel of pyEfis.
In any case, I think this webview instrument is handy to display other thing too. Someone (like @e100) who knows the pyEfis well could do more polished version of this instrument!
![20250121_183441](https://github.com/user-attachments/assets/5d351717-c858-4469-8db4-5a22328eb48a)
